### PR TITLE
Annotations: Add typeahead support for tags in built-in annotations

### PR DESCRIPTION
**Pull request by ashharrison90**
**What this PR does / why we need it**:

- exposes new props on `TagFilter`:
  - `allowCustomValue` to allow for optional creation of tags
  - `inputId` to hook up labels correctly in `<InlineField>`s
- adds a default count of 0 to the `TagBadge` in `TagOption` to prevent showing `undefined` when creating a tag
- adds the new backend route to get the list of annotation tags and their counts
- hides the `AnnotationFieldMapper` for the built in datasource (lmk if there's a better way to implement this!)
- creates a new React component for the built-in annotations with some basic unit tests
- modifies the grafana datasource to work with this new component
- removes the old angular code! 🎉 

**Which issue(s) this PR fixes**:

For https://github.com/grafana/grafana/issues/36107

**Special notes for your reviewer**:

- modifies the grafana datasource to work with this new component - **this is the bit i'm most unsure about as the typings/interface seems super complicated.** 😕

### CODE REVIEW

File: public/app/plugins/datasource/grafana/components/AnnotationQueryEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    A tag entered here as {"'"}foo{"'"} will match
+    <ul>
+      <li>
+        annotation tags {"'"}foo{"'"}
+      </li>
+      <li>
+        annotation key-value tags formatted as {"'"}foo:bar{"'"}
+      </li>
+    </ul>
+  </div>
```
Comment (dprokop) [inline public/app/plugins/datasource/grafana/components/AnnotationQueryEditor.tsx:34]: This produces weirdly formatted list in the UI:

image

**Review by dprokop — COMMENTED** [review]
Took this for a spin, tested and quickly reviewed and all seems good :)

File: public/app/plugins/datasource/grafana/components/AnnotationQueryEditor.tsx (hunk 1 above)
Comment (ashharrison90) [reply public/app/plugins/datasource/grafana/components/AnnotationQueryEditor.tsx:34]: ah yes, i noticed that and forgot to do anything about it, thanks! 😂 
we override the padding on all `ul`/`ol` here: https://github.com/grafana/grafana/blob/main/public/sass/base/_type.scss#L50

we can remove that, but i'm not sure what else it affects 😕  it must have been added for a reason

File: public/app/plugins/datasource/grafana/components/AnnotationQueryEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  { label: 'Tags', value: GrafanaAnnotationType.Tags },
+];
+
+const limitOptions = [10, 50, 100, 200, 300, 500, 1000, 2000].map((limit) => ({
+  label: String(limit),
+  value: limit,
+}));
+
+interface Props {
+  query: GrafanaQuery;
```
Comment (ashharrison90) [inline public/app/plugins/datasource/grafana/components/AnnotationQueryEditor.tsx:48]: really this should be typed as `GrafanaAnnotationQuery`, but doing this creates type errors in `datasource`

File: public/app/plugins/datasource/grafana/datasource.ts [hunk 3]
```diff
@@ -8,24 +11,51 @@ import {
...
+        // This should handle migrating them
+        json.target = json.target ?? {
+          type: json.type ?? GrafanaAnnotationType.Dashboard,
+          limit: json.limit ?? 100,
+          tags: json.tags ?? [],
+          matchAny: json.matchAny ?? false,
+        }; // using spread syntax caused an infinite loop in StandardAnnotationQueryEditor
+        return json;
+      },
+      prepareQuery(anno: AnnotationQuery<GrafanaAnnotationQuery>): GrafanaQuery {
```
Comment (ashharrison90) [inline public/app/plugins/datasource/grafana/datasource.ts:39]: i'm not sure this is really a `GrafanaQuery`, but it doesn't seem to complain. this is basically `AnnotationQuery<GrafanaAnnotationQuery>` with a `queryType` and `refId`

File: public/app/plugins/datasource/grafana/datasource.ts [hunk 4]
```diff
@@ -80,30 +110,31 @@ export class GrafanaDatasource extends DataSourceApi<GrafanaQuery> {
     return Promise.resolve([]);
   }
 
-  annotationQuery(options: AnnotationQueryRequest<GrafanaQuery>): Promise<AnnotationEvent[]> {
+  async getAnnotations(options: AnnotationQueryRequest<GrafanaQuery>): Promise<DataQueryResponse> {
     const templateSrv = getTemplateSrv();
-    const annotation = (options.annotation as unknown) as GrafanaAnnotationQuery;
+    const annotation = (options.annotation as unknown) as AnnotationQuery<GrafanaAnnotationQuery>;
```
Comment (ashharrison90) [inline public/app/plugins/datasource/grafana/datasource.ts:115]: have to cast the annotation here

**Review by ashharrison90 — COMMENTED** [review]
typing concerns

**Comment by ashharrison90** [comment]
also copying my comment from slack so it's all in one place:
> there was also a couple of questions that came out of our chat with docs:
> - what graph types support the in built annotations? is it only graph(old) and timeseries?
> - the docs here (https://grafana.com/docs/grafana/latest/dashboards/annotations/#query-by-tag) mention that you can use template variables in the tag query. currently in the PR you can't do this, as the query is restricted to only allow adding one of the options passed to TagFilter which are the currently defined tags. i think there are 2 options for enabling:
>   - allow adding custom values in this input. this is a 1-line change
>   - merge any template variables into the list of options passed to TagFilter . Is there an API route where we can query what template variables there are? 🤔 
> 
> - my slight preference would be option 2, but option 1 is an easy fallback and arguably gives more flexibility. (edited)

**Comment by torkelo** [comment]
@ashharrison90 awesome job on this! 
@dprokop  

What do you thin of switching to new form styles for this? Moving tooltips to the description. And for the filter by I moved the description of each mode to the select item description. 

Screenshot from 2021-07-05 15-59-14

**Comment by ashharrison90** [comment]
yeah, i prefer that. removes all the hover tooltips which is better from an accessibility point of view as well. 🥳

**Comment by torkelo** [comment]
@ashharrison90 exactly. feel free to apply the above patch, paste in a file named patch (paste it in a file and  `git apply patch`)

**Comment by torkelo** [comment]
@ashharrison90 or might be easier for me just just push those, one sec

**Comment by dprokop** [comment]
I'm all -in when it comes to forms updates :)

**Comment by ashharrison90** [comment]
@dprokop @torkelo marked this ready for review now i think we've made all the necessary changes.

i've pushed a change to allow custom values in the `AnnotationQueryEditor`. this will preserve being able to query tags by a template variable for now. if there is an api to call to get the list of defined template variables, i can call that instead and merge them with the tag options.

**Comment by torkelo** [comment]
>  if there is an api to call to get the list of defined template variables, i can call that instead and merge them with the tag options.

Don't think that is needed for now.

**Review by torkelo — APPROVED** [review]

**Pull request merged**